// aws-cdk/test/api/hotswap-deployments.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const hotswap_deployments_1 = require("../../lib/api/hotswap-deployments");
const util_1 = require("../util");
const mock_sdk_1 = require("../util/mock-sdk");
const fake_cloudformation_stack_1 = require("./fake-cloudformation-stack");
const STACK_NAME = 'withouterrors';
const STACK_ID = 'stackId';
let mockSdkProvider;
let mockUpdateLambdaCode;
let currentCfnStack;
beforeEach(() => {
    jest.resetAllMocks();
    mockSdkProvider = new mock_sdk_1.MockSdkProvider({ realSdk: false });
    mockUpdateLambdaCode = jest.fn();
    mockSdkProvider.stubLambda({
        updateFunctionCode: mockUpdateLambdaCode,
    });
    currentCfnStack = new fake_cloudformation_stack_1.FakeCloudformationStack({
        stackName: STACK_NAME,
        stackId: STACK_ID,
    });
});
test('returns a deployStackResult with noOp=true when it receives an empty set of changes', async () => {
    // WHEN
    const deployStackResult = await hotswap_deployments_1.tryHotswapDeployment(mockSdkProvider, {}, currentCfnStack, cdkStackArtifactOf());
    // THEN
    expect(deployStackResult).not.toBeUndefined();
    expect(deployStackResult === null || deployStackResult === void 0 ? void 0 : deployStackResult.noOp).toBeTruthy();
    expect(deployStackResult === null || deployStackResult === void 0 ? void 0 : deployStackResult.stackArn).toEqual(STACK_ID);
});
test('returns undefined when it a new Lambda function is added to the Stack', async () => {
    // GIVEN
    const cdkStackArtifact = cdkStackArtifactOf({
        template: {
            Resources: {
                Func: {
                    Type: 'AWS::Lambda::Function',
                },
            },
        },
    });
    // WHEN
    const deployStackResult = await hotswap_deployments_1.tryHotswapDeployment(mockSdkProvider, {}, currentCfnStack, cdkStackArtifact);
    // THEN
    expect(deployStackResult).toBeUndefined();
});
test('calls the updateLambdaCode() API when it receives only a code difference in a Lambda function', async () => {
    // GIVEN
    currentCfnStack.setTemplate({
        Resources: {
            Func: {
                Type: 'AWS::Lambda::Function',
                Properties: {
                    Code: {
                        S3Bucket: 'current-bucket',
                        S3Key: 'current-key',
                    },
                    FunctionName: 'my-function',
                },
                Metadata: {
                    'aws:asset:path': 'old-path',
                },
            },
        },
    });
    const cdkStackArtifact = cdkStackArtifactOf({
        template: {
            Resources: {
                Func: {
                    Type: 'AWS::Lambda::Function',
                    Properties: {
                        Code: {
                            S3Bucket: 'current-bucket',
                            S3Key: 'new-key',
                        },
                        FunctionName: 'my-function',
                    },
                    Metadata: {
                        'aws:asset:path': 'new-path',
                    },
                },
            },
        },
    });
    // WHEN
    const deployStackResult = await hotswap_deployments_1.tryHotswapDeployment(mockSdkProvider, {}, currentCfnStack, cdkStackArtifact);
    // THEN
    expect(deployStackResult).not.toBeUndefined();
    expect(mockUpdateLambdaCode).toHaveBeenCalledWith({
        FunctionName: 'my-function',
        S3Bucket: 'current-bucket',
        S3Key: 'new-key',
    });
});
function cdkStackArtifactOf(testStackArtifact = {}) {
    return util_1.testStack({
        stackName: STACK_NAME,
        ...testStackArtifact,
    });
}
