// aws-cdk/lib/api/util/cloudformation/evaluate-cfn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.evaluateCfn = void 0;
function evaluateCfn(object, context) {
    const intrinsicFns = {
        'Fn::Join'(separator, args) {
            return evaluate(args).map(evaluate).join(separator);
        },
        'Fn::Split'(separator, args) {
            return evaluate(args).split(separator);
        },
        'Fn::Select'(index, args) {
            return evaluate(args).map(evaluate)[index];
        },
        'Ref'(logicalId) {
            if (logicalId in context) {
                return context[logicalId];
            }
            else {
                throw new Error(`Reference target '${logicalId}' was not found`);
            }
        },
        'Fn::Sub'(template, explicitPlaceholders) {
            const placeholders = explicitPlaceholders
                ? { ...context, ...evaluate(explicitPlaceholders) }
                : context;
            return template.replace(/\${([^}]*)}/g, (_, key) => {
                if (key in placeholders) {
                    return placeholders[key];
                }
                else {
                    throw new Error(`Fn::Sub target '${key}' was not found`);
                }
            });
        },
    };
    return evaluate(object);
    function evaluate(obj) {
        if (Array.isArray(obj)) {
            return obj.map(evaluate);
        }
        if (typeof obj === 'object') {
            const intrinsic = parseIntrinsic(obj);
            if (intrinsic) {
                return evaluateIntrinsic(intrinsic);
            }
            const ret = {};
            for (const key of Object.keys(obj)) {
                ret[key] = evaluate(obj[key]);
            }
            return ret;
        }
        return obj;
    }
    function evaluateIntrinsic(intrinsic) {
        if (!(intrinsic.name in intrinsicFns)) {
            throw new Error(`Intrinsic ${intrinsic.name} not supported here`);
        }
        const argsAsArray = Array.isArray(intrinsic.args) ? intrinsic.args : [intrinsic.args];
        return intrinsicFns[intrinsic.name].apply(intrinsicFns, argsAsArray);
    }
}
exports.evaluateCfn = evaluateCfn;
function parseIntrinsic(x) {
    if (typeof x !== 'object' || x === null) {
        return undefined;
    }
    const keys = Object.keys(x);
    if (keys.length === 1 && (keys[0].startsWith('Fn::') || keys[0] === 'Ref')) {
        return {
            name: keys[0],
            args: x[keys[0]],
        };
    }
    return undefined;
}
